File: predict-mnist.cs
using System;
using System.Collections.Generic;
using System.IO;

namespace MNISTClassifier
{
    class Program
    {
        static int Square(int x)
        {
            return x * x;
        }

        static int SumSquaredError(List<int> x, List<int> y, int start)
        {
            if (x.Count != y.Count)
            {
                throw new ArgumentException("Lists must have the same size.");
            }

            int sum = 0;
            for (int i = start; i < x.Count; i++)
            {
                sum += Square(x[i] - y[i]);
            }

            return sum;
        }

        static void Main(string[] args)
        {
            Console.WriteLine("!");

            const string fileNameTrain = "mnist_train.csv";
            const string fileNameTest = "mnist_test.csv";

            // Load train data
            var trainData = new List<List<int>>();
            using (var reader = new StreamReader(fileNameTrain))
            {
                while (!reader.EndOfStream)
                {
                    var line = reader.ReadLine();
                    var values = line.Split(',');

                    var data = new List<int>();
                    foreach (var value in values)
                    {
                        data.Add(int.Parse(value));
                    }

                    trainData.Add(data);
                }
            }
            Console.WriteLine($"Train data loaded. rows: {trainData.Count}");

            // Test
            using (var reader = new StreamReader(fileNameTest))
            {
                int nRows = 0;
                int nCorrect = 0;

                while (!reader.EndOfStream)
                {
                    nRows++;

                    var line = reader.ReadLine();
                    var values = line.Split(',');

                    var testData = new List<int>();
                    foreach (var value in values)
                    {
                        testData.Add(int.Parse(value));
                    }

                    // Predict
                    int best = 0;
                    int minError = -1;
                    foreach (var data in trainData)
                    {
                        var error = SumSquaredError(testData, data, 1);
                        if (minError < 0 || error < minError)
                        {
                            minError = error;
                            best = data[0];
                        }
                    }

                    if (best == testData[0])
                    {
                        nCorrect++;
                    }

                    Console.WriteLine($"row: {nRows} predicted: {best} answer: {testData[0]} accuracy: {(double)nCorrect / nRows}");
                }
            }
        }
    }
}

File: predict-mnist.jl
using DelimitedFiles
using Random

square(x) = x * x

function sum_sq_err(xs::Vector, ys::Vector, start::Int)
    s = 0
    for i in start:length(xs)
        s += square(xs[i] - ys[i])
    end
    return s
end

function predict(train_data::Vector{Vector{Float64}}, test::Vector{Float64})
    predicted = 0.0
    min_error = -1.0    
    for train in train_data
        error = sum_sq_err(test, train, 1)
        if error < min_error || min_error < 0
            predicted = train[1]
            min_error = error
        end
    end
    return predicted
end

function main()
    file_name_train = "mnist_train.csv"
    file_name_test = "mnist_test.csv"

    # train
    train_data = open(file_name_train) do io
        lines = readlines(io)
        [parse.(Float64, split(line, ",")) for line in lines]
    end
    println("train data loaded. rows: $(length(train_data))")

    # test
    n_rows = 0
    n_correct = 0
    open(file_name_test) do io
        for line in eachline(io)
            test = parse.(Float64, split(line, ","))
            n_rows += 1
            predicted = predict(train_data, test)
            answer = test[1]
            if predicted == answer
                n_correct += 1
            end
            accuracy = n_correct / n_rows
            println("row: $n_rows predicted: $predicted answer: $answer accuracy: $accuracy")
        end
    end
end

main()